src/gedi_subset/maapx.py

```diff
@@ -2,109 +2,188 @@
+logger = logging.getLogger(__name__)
 
-logger = logging.getLogger(f"gedi_subset.{__name__}")
 
+__all__ = [
+    "AWSCredentials",
+    "download_granule",
+    "find_collection",
+]
 
-def _is_s3_credentials_online_resource(resource) -> bool:
+T = TypeVar("T")
+P = ParamSpec("P")
+
+
+class AWSCredentials(TypedDict):
+    accessKeyId: str
+    secretAccessKey: str
+    sessionToken: str
+
+
+def _is_s3_credentials_online_resource(resource: Mapping[str, str]) -> bool:
+    """Determine whether or not a granule's online resource is an S3 credentials
+    endpoint.
+    """
     url = resource.get("URL", "").lower()
     description = resource.get("Description", "").lower()
 
     # TODO: is this sufficient for identifying URL for obtaining S3 credentials?
     return url and url.endswith("/s3credentials") or "credentials" in description
 
 
-def _s3_credentials_endpoint(granule: Granule) -> Maybe[str]:
-    return flow(
-        granule.get("Granule", {}).get("OnlineResources", {}).get("OnlineResource", []),
-        lambda resources: resources if isinstance(resources, list) else [resources],
-        fp.find(_is_s3_credentials_online_resource),
-        bind(safe(operator.itemgetter("URL"))),
-        result_to_maybe,
-    )
+def _s3_credentials_endpoint(granule: Granule) -> Optional[str]:
+    """Return the S3 credentials endpoint for a granule.
+
+    Parameters
+    ----------
+    granule
+        The granule for which to obtain the S3 credentials endpoint.
+
+    Returns
+    -------
+    url
+        The S3 credentials endpoint URL or `None` if no S3 credentials endpoint
+        is found.
+    """
+    granule = granule.get("Granule", {})
+    resources = granule.get("OnlineResources", {}).get("OnlineResource", [])
+    # OnlineResources.OnlineResource is a list only if there are multiple
+    # resources, so when there is only one resource, we must put it in a list.
+    # This is due to how the XML is parsed by the `xmltodict` library.
+    resources = resources if isinstance(resources, list) else [resources]
+
+    # We assume there is only one S3 credentials endpoint for a granule.
+    if s3_resource := next(filter(_is_s3_credentials_online_resource, resources), None):
+        return s3_resource["URL"]
+
+    return None
 
 
-@ttl_cache(ttl=55 * 60)
-def _earthdata_s3_credentials(maap: MAAP, endpoint: str) -> IOResultE["AWSCredentials"]:
-    """Returns short-term AWS credentials obtained from an S3 credentials endpoint."""
+@cached(cache=TTLCache(maxsize=1, ttl=55 * 60), key=lambda _, endpoint: endpoint)
+def _s3_credentials(maap: MAAP, endpoint: str) -> AWSCredentials | Never:
+    """Obtain short-term AWS credentials from an S3 credentials endpoint."""
 
     logger.debug(f"Obtaining S3 credentials from {endpoint}")
 
-    return impure_safe(maap.aws.earthdata_s3_credentials)(endpoint)
+    return maap.aws.earthdata_s3_credentials(endpoint)
 
 
 @cached(cache=FIFOCache(maxsize=1), key=lambda creds: creds["sessionToken"])
```

Comment:
Good question. In part, the answer is that this caching logic predates any other caching that may now be implemented on the server side.

However, the other caching you're referring to, if my understanding is correct, occurs on the MAAP API side, certainly not within the `maap-py` client side. So while in the end, the caching on the API server side ultimately prevents unnecessary token requests (if indeed that's what is implemented -- I haven't looked at the API code in quite some time), the caching done here at least prevents unnecessary network traffic.

It might indeed now be safe to eliminate the caching here, but for the moment, I'd prefer to leave it in place. I do have an alternative to this, which I plan to address in #14, but I don't want to further complicate this PR with such changes.